### PR TITLE
Proxy handler

### PR DESCRIPTION
**Pull request by neosbob**
Hello, this is TeamHello.

I am creating this pull request to keep track of all my changes for the proxy handler.

Progress:
    -Added files for proxy_handler.
    -updated config file and implemented Init Function for proxy_handler.
    -Added Integration test for 302 and backend proxy.
   -Polished up unit-tests.
   -Finished implementing proxy_handler.

TODO:
    -Integration test.
    -More unit tests.

Do Not Merge.

### CODE REVIEW

**Comment by RBarank** [comment]
Hey just let us know when you're ready to merge.

**Comment by neosbob** [comment]
There seems to be an issue where the server dumps core/ segmentation fault.
This happens when the browser is trying to request additional files from proxy server.

I suspect that this is because in connection.cc, the server doesn't know to use the proxy handler set with the path "/".  Currently, connection can't find any matching request handler to handle the subsequent requests.

The issue lies within the  second_slash search method.  The URI "/css/main.css" assumes the request_path is "/css", and doesn't check to see if any existing handler can handle the URI.

**Comment by abineetds** [comment]
There is a bug with the uri_prefix being "/". That does dump core, but it shouldn't hamper anything unless you specify a handler with a uri_prefix of "/" in the config.

The problem you are talking about of the uri /css/main.css. The code will treat /css to be a uri prefix and try to search for a handler associated with that. You're right there is a bug that it ends up dumping core if the uri_prefix is not found in the map. That's a quick fix, to just search the map before trying to access it and set a bad request if not found. 

Would you like us to fix that, or would you? I don't want there to be a merge conflict.

**Comment by neosbob** [comment]
Actually it does this after  saving a value to request_base: 
     RequestHandler::Status ret = pathMap_[request_base]->HandleRequest(*currentRequest, resp);

So if a request_base is not found, it does some weird stuff. Probably due to accessing nullptr.

Should you guys handle the fix? Or do I apply a fix? Not sure on this.

**Comment by abineetds** [comment]
I know exactly what the problem is. It's only a small check that needs to
be done. I'm not on a machine right now, I can do it in the afternoon
tomorrow. It's only four or five lines of code, do its your choice. If I
fix it you're going to have to pull again.

**Comment by echen94** [comment]
There seems to be some issue in the Parse function of request.cc. 
There is an iterator error when we tried the Parse function for a request in our unit tests.
Also, the server crashes when we send another request during one session. It should not be a problem from the proxy server because it works fine with our own code. Let us know if you can figure out where the problem comes from. 

Here is the error message we got:
const Type& boost::token_iterator<TokenizerFunc, Iterator, Type>::dereference() const [with TokenizerFunc = boost::char_separator<char>; Iterator = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; Type = std::__cxx11::basic_string<char>]: Assertion `valid_' failed.

**Comment by RBarank** [comment]
Ok we will look into it

**Comment by abineetds** [comment]
Does the server crash have something to do with /favicon.ico?

**Comment by abineetds** [comment]
What was the request you were sending when you got the iterator error?

**Comment by neosbob** [comment]
So I open a session, enter localhost:3000/ in the browser. Everything works fine, the ucla webpage is displayed. 

Same session, enter localhost:3000/ in the browser. The iterator error occurs, and server crashes.

The favicon does cause the server to crash on localhost:3000/echo.

**Comment by abineetds** [comment]
That's odd, let me check

**Comment by abineetds** [comment]
I ran your most recent code, didn't give me an issue. The only thing I changed was I commented a cout statement in your proxy handler that was really populating the console. cout.write()

**Comment by neosbob** [comment]
Try entering localhost:3000/ in the browser multiple times in one session.

**Comment by neosbob** [comment]
The error pops up sporadically when certain conditions have been fulfilled.  I have also tested this on my team's server, and the above mentioned error did not appear.

**Comment by abineetds** [comment]
That's weird, I think it might be an issue in which we cut off a request to
start sending fresh ones. I didn't see this problem but maybe the
multithreading will fix this.

**Comment by neosbob** [comment]
So minus the above mentioned error, everything else works fine. Is it ok if we merge later?

**Comment by neosbob** [comment]
Ready for merging.

**Comment by neosbob** [comment]
Need an approved review.

File: proxy_handler.cc [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    std::getline(ssHeader, value, '\r');
+    
+    std::pair<std::string, std::string> header_pair(headerName, value);
+    this->headers_.push_back(header_pair);
+  }
+  return 1;
+	
+}
+
+bool ProxyHandler::url_decode(const std::string& in, std::string& out)
```
Comment (panjiyar) [inline proxy_handler.cc]: is this function being used anywhere?

**Review by panjiyar — CHANGES_REQUESTED** [review]
Do you need the url_decode function in the handler. i don't see it being used

**Comment by neosbob** [comment]
Oh ok, I'll remove it.

**Comment by neosbob** [comment]
Removed the function.

**Review by panjiyar — APPROVED** [review]

**Comment by panjiyar** [comment]
Will merge once check completes

**Comment by neosbob** [comment]
Ok. Thanks!

**Comment by abineetds** [comment]
Good job guys!

**Pull request merged**